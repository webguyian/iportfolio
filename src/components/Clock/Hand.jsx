import React from 'react';
import PropTypes from 'prop-types';

const Hand = props => {
  const { angle, length, width, ...otherProps } = props;

  return (
    <line
      y2={length}
      transform={`rotate(${angle} ${width} ${width})`}
      {...otherProps}
    />
  );
};

Hand.propTypes = {
  angle: PropTypes.number,
  className: PropTypes.string,
  x1: PropTypes.number,
  x2: PropTypes.number,
  y1: PropTypes.number,
  length: PropTypes.number,
  width: PropTypes.number
};

Hand.defaultProps = {
  x1: 30,
  x2: 30,
  y1: 30,
  length: 60,
  width: 30
};

export default Hand;
