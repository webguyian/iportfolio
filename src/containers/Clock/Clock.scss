.clock-app,
.stopwatch {
  @include stretch-column;
}

.stopwatch {
  flex-grow: 1;
  align-items: center;
  justify-content: center;
  height: auto;

  &-text {
    @include type-number-display;
    margin: auto 0 100px;
    text-align: center;
  }
}
