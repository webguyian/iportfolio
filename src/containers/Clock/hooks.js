import { useEffect, useRef, useState } from 'react';
import { useHistory } from 'react-router-dom';

import { initialValues } from './constants';
import { getExpiration, getExpirationValues, getSeconds } from './helpers';

export const useInitialRoute = (match, path) => {
  const history = useHistory();

  useEffect(() => {
    if (match.isExact) {
      history.push(path);
    }
  }, []);

  return history;
};

export const useStorage = (key, currentValue, deleteKey, keepUpdating) => {
  const getValue = () => {
    try {
      // Get item from local storage with key
      const item = window.localStorage.getItem(key);

      // Parse stored JSON or if none return undefined
      return item ? JSON.parse(item) : undefined;
    } catch (error) {
      // If error return undefined
      return undefined;
    }
  };

  const value = getValue();
  const localRef = useRef(value);
  const [storedValue, setStoredValue] = useState(value);

  const setValue = nextValue => {
    try {
      // Save to state and local storage
      setStoredValue(nextValue);

      if (deleteKey && !nextValue[deleteKey]) {
        // Remove item from local storage
        window.localStorage.removeItem(key);
      } else {
        // Update local storage and include timestamp
        nextValue.timestamp = Date.now();
        window.localStorage.setItem(key, JSON.stringify(nextValue));
      }
    } catch (error) {
      // Drop error
    }
  };

  useEffect(() => {
    // Update local value
    localRef.current = currentValue || storedValue;

    if (keepUpdating) {
      // Set update value in local storage
      setValue(localRef.current);
    }
  }, [currentValue, storedValue]);

  useEffect(() => {
    return () => {
      if (key && localRef.current) {
        setValue(localRef.current);
      }
    };
  }, []);

  return localRef.current;
};

export const useLaps = (initialLaps, timer, timerRunning) => {
  const [laps, setLaps] = useState(initialLaps);
  const hasLaps = laps.some(lap => lap > 0);
  const storage = useStorage(
    'stopwatch',
    {
      laps,
      timer,
      timerRunning
    },
    'timer'
  );

  const updateLaps = () => {
    if (timerRunning) {
      const hasEmptyLaps = laps.some(lap => lap === 0);

      if (hasEmptyLaps) {
        setLaps(updatedLaps => [
          0,
          ...updatedLaps.slice(0, updatedLaps.length - 1)
        ]);
      } else {
        setLaps(updatedLaps => [0, ...updatedLaps]);
      }
    } else {
      setLaps(initialLaps);
    }
  };

  useEffect(() => {
    if (storage && storage.laps) {
      setLaps(storage.laps);
    }
  }, []);

  useEffect(() => {
    if (!timerRunning || !timer) {
      return;
    }

    const lapLength = laps.length;

    if (!hasLaps && timer > 10) {
      setLaps(updatedLaps => [timer, ...updatedLaps]);
    }

    if (lapLength > 5) {
      setLaps(updatedLaps => {
        const truncatedLaps = updatedLaps.slice(1, lapLength);
        const offset = truncatedLaps.reduce((a, b) => a + b, 0);

        return [timer - offset, ...truncatedLaps];
      });
    }
  }, [timer, timerRunning, hasLaps]);

  return [laps, updateLaps, hasLaps];
};

export const useStopwatch = (startTime = 0, start = false) => {
  const [timerStarted, startTimer] = useState(start);
  const [interval, setLocalInterval] = useState(null);
  const [offset, setOffset] = useState(0);
  const timer = useRef(startTime);
  const storage = useStorage('stopwatch');
  const toggleTimer = () => startTimer(!timerStarted);
  const resetTimer = () => {
    timer.current = 0;
    return timer.current;
  };
  const getUpdatedTime = cache => {
    const now = Date.now();
    const diff = now - cache.timestamp;

    return cache.timer + diff;
  };

  useEffect(() => {
    if (storage && storage.timer) {
      // Get updated time from local storage
      if (storage.timerRunning) {
        timer.current = getUpdatedTime(storage);
        toggleTimer();
      } else {
        timer.current = storage.timer;
      }
    }
  }, []);

  useEffect(() => {
    let localOffset = offset;

    if (timerStarted && !interval) {
      localOffset = Date.now();

      const getDelta = () => {
        const now = Date.now();
        const delta = now - localOffset;

        localOffset = now;

        setOffset(now);
        return delta;
      };

      const localInterval = window.setInterval(() => {
        timer.current += getDelta();
      }, 1);

      setLocalInterval(localInterval);
    }

    if (!timerStarted && interval) {
      window.clearInterval(interval);
      setLocalInterval(null);
    }
  }, [timerStarted, timer, interval]);

  return [timer.current, timerStarted, toggleTimer, resetTimer];
};

export const useDuration = () => {
  const [values, setValues] = useState(null);
  const storage = useStorage('duration', values, null, true);
  const setDuration = (key, value) => {
    setValues({
      ...values,
      [key]: value
    });
  };
  const cachedValues = storage || values;
  const duration = cachedValues && getSeconds(cachedValues);

  useEffect(() => {
    if (storage) {
      // Update state with values from local storage
      setValues(storage);
    } else {
      // Set default initial values
      setValues(initialValues);
    }
  }, []);

  return [duration, setDuration, cachedValues, setValues];
};

export const useCountdown = (h = 0, m = 0, s = 0) => {
  const [seconds, setSeconds] = useState(s);
  const [minutes, setMinutes] = useState(m);
  const [hours, setHours] = useState(h);
  const allSeconds = getSeconds({ hours, minutes, seconds });

  const subtractHour = () => {
    setHours(prevHours => {
      if (prevHours === 0) {
        return 23;
      } else if (prevHours > 0) {
        return prevHours - 1;
      }
      return 0;
    });
  };

  const subtractMinute = () => {
    setMinutes(prevMinutes => {
      if (prevMinutes === 0) {
        subtractHour();
        return 59;
      } else if (prevMinutes > 0) {
        return prevMinutes - 1;
      }
      return 0;
    });
  };

  const subtractSecond = () => {
    setSeconds(prevSeconds => {
      if (prevSeconds === 0) {
        subtractMinute();
        return 59;
      } else if (prevSeconds > 0) {
        return prevSeconds - 1;
      }
      return 0;
    });
  };

  const setCountdown = values => {
    if (!values) {
      setHours(0);
      setMinutes(0);
      setSeconds(0);
      return;
    }

    setHours(values.hours);
    setMinutes(values.minutes);
    setSeconds(values.seconds);
  };

  return [
    {
      allSeconds,
      hours,
      minutes,
      seconds
    },
    setCountdown,
    subtractSecond
  ];
};

export const useTimer = duration => {
  const [started, setStarted] = useState(false);
  const [running, setRunning] = useState(false);
  const [countdown, setCountdown, startCountdown] = useCountdown();
  const intervalRef = useRef();
  const storage = useStorage(
    'timer',
    {
      ...countdown,
      duration,
      running,
      started
    },
    'allSeconds'
  );

  const pause = () => {
    if (intervalRef.current) {
      setRunning(() => false);
      clearInterval(intervalRef.current);
      intervalRef.current = undefined;
    }
  };

  const reset = () => {
    pause();
    setCountdown(null);
    setStarted(false);
  };

  const calculateExpiration = expiration => {
    const distance = expiration - Date.now();
    const expirationValues = getExpirationValues(distance);

    if (countdown.allSeconds < 0) {
      reset();
    } else {
      setCountdown(expirationValues);
    }

    return expirationValues;
  };

  const start = expiration => {
    if (expiration && !intervalRef.current) {
      calculateExpiration(expiration);
    }
  };

  const resume = () => {
    if (!intervalRef.current) {
      setStarted(() => true);
      setRunning(() => true);
      intervalRef.current = setInterval(startCountdown, 1000);
    }
  };

  const restart = () => {
    const expiration = getExpiration(duration);

    reset();
    start(expiration);
  };

  const toggle = () => {
    if (!intervalRef.current) {
      resume();
    } else {
      pause();
    }
  };

  useEffect(() => {
    const expiration = getExpiration(duration, storage);

    start(expiration);

    if (storage) {
      if (storage.running) {
        resume();
      } else if (storage.started) {
        setStarted(true);
      }

      if (duration !== storage.duration) {
        // Remove timer from local storage
        window.localStorage.removeItem('timer');
      }
    }

    return pause;
  }, [duration]);

  return {
    ...countdown,
    duration,
    start,
    started,
    pause,
    reset,
    resume,
    restart,
    running,
    toggle
  };
};

export const useRadialOffset = (timer, radius) => {
  const circumference = (2 * radius * 22) / 7;
  const { allSeconds, duration } = timer;
  const [initialSeconds] = useState(allSeconds);
  const [offset, setOffset] = useState(0);

  useEffect(() => {
    if (!offset) {
      const initialOffset = (allSeconds / duration) * circumference;
      const updatedOffset = circumference - Math.round(initialOffset);

      if (updatedOffset === circumference || updatedOffset < 0) {
        // Exit early
        return;
      }

      setOffset(updatedOffset);
    }
  }, [allSeconds, duration, offset]);

  return [offset, initialSeconds];
};
