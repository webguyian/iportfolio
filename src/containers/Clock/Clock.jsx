import React from 'react';
import PropTypes from 'prop-types';

import TabNavigation from 'components/TabNavigation/TabNavigation';
import Stopwatch from 'containers/Clock/Stopwatch/Stopwatch';
import Timer from 'containers/Clock/Timer/Timer';

import { useInitialRoute } from './hooks';

const Clock = props => {
  const { match } = props;
  const baseClass = 'clock-app';
  const tabs = [
    {
      component: Stopwatch,
      icon: 'stopwatch',
      label: 'Stopwatch',
      path: `${match.path}/stopwatch`
    },
    {
      component: Timer,
      icon: 'clock',
      label: 'Timer',
      path: `${match.path}/timer`
    }
  ];

  useInitialRoute(match, tabs[0].path);

  return (
    <div className={baseClass}>
      <TabNavigation tabs={tabs} />
    </div>
  );
};

Clock.propTypes = {
  match: PropTypes.object.isRequired
};

export default Clock;
