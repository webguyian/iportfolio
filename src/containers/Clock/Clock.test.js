import React from 'react';
import { create } from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';

import Clock from './Clock';

describe('<Clock />', () => {
  const props = {
    match: {
      path: '/clock'
    }
  };

  it('renders correctly', () => {
    const component = create(
      <MemoryRouter>
        <Clock {...props} />
      </MemoryRouter>
    );

    expect(component).toMatchSnapshot();
  });
});
