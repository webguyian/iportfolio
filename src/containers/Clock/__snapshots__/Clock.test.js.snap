// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Clock /> renders correctly 1`] = `
<div
  className="clock-app"
>
  <nav
    className="ui-tab-navigation"
  >
    <ul
      className="ui-tab-navigation-list"
    >
      <li
        className="ui-tab-navigation-list-item"
      >
        <a
          aria-current={null}
          className="ui-tab-navigation-link ui-link"
          href="/clock/stopwatch"
          onClick={[Function]}
        >
          <i
            className="ui-icon ui-icon-stopwatch fas fa-stopwatch fa-3x"
          />
          <span
            className="ui-text ui-text--body ui-tab-navigation-link-label"
          >
            Stopwatch
          </span>
        </a>
      </li>
      <li
        className="ui-tab-navigation-list-item"
      >
        <a
          aria-current={null}
          className="ui-tab-navigation-link ui-link"
          href="/clock/timer"
          onClick={[Function]}
        >
          <i
            className="ui-icon ui-icon-clock fas fa-clock fa-3x"
          />
          <span
            className="ui-text ui-text--body ui-tab-navigation-link-label"
          >
            Timer
          </span>
        </a>
      </li>
    </ul>
  </nav>
</div>
`;
